docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+
+ - Foreign code should not imply unsafety.
+ - Interoperability should not compromise any security guarantees of the non-Verona language or library API.
+ - Most libraries are stateful, the correct granularity for interoperability is a library, not a function.
+
+The first of these relates to one of the core principles of Verona: there is no `unsafe` keyword.
+Verona targets security-critical workloads and it's very hard to make strong security claims about a language if any call to another language is able to compromise any and all of our safety guarantees.
+The Java JNI specification, for example, contains this disclaimer in the Design Overview chapter:
+
+> The programmer must not pass illegal pointers or arguments of the wrong type to JNI functions. Doing so could result in arbitrary consequences, including a corrupted system state or VM crash.
+
+The language makes an explicit declaration that any call to native code can undermine any of the invariants that the Java security model depends on. 
+This is a pragmatic decision to allow quick adoption but it means that no Java program can be considered safe until all native components have been removed.
+
+Modern C++ codebases work hard to provide memory safety guarantees at the API level.
+These guarantees are often lost in interop layers.
+For example, LLVM makes extensive use of typed smart pointers in the C++ APIs, but the Python bindings are written on top of the C API, which make type safety the caller's responsibility.
+Even if we can sandbox the resulting code, using a C++ library from Verona should not be more likely to introduce security vulnerabilities than using the same library from C++.
+The 2017 paper [*Finding and Preventing Bugs in JavaScript Bindings*](https://ieeexplore.ieee.org/document/7958598) showed that memory safety bugs in the C++ / JavaScript interop layer in Chrome are surprisingly common.
+Any time that a user has to hand-write code at a bindings layer to get memory safety right, we can assume that they will get it wrong and so we aim to automate as much as possible.
+
+The desire for sandboxing also leads to the conclusion that a 'foreign function interface' is exposing the wrong level of abstraction.
+What sandbox should a function run in?
+Our prior [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work showed that treating libraries (in the loose sense, including any dependent libraries) as the unit of foreign code provided an abstraction that fitted well with sandboxing and allowed the user to determine the degree of state sharing between foreign code invocations.
+This abstraction also has software engineering benefits.
+LLVM, for example, does not have stable ABI (or API) guarantees across programs and so can cause linking errors if, for example, an application depends on one library that uses libclang from LLVM 10 to parse syntax highlighting and another that uses the JIT from LLVM 11 for GPU shaders.
```

Comment:
I'm not sure I understand this comment.

Reply:
If the errors are not a link time but, say, use of the wrong enum values and is only observable at run time. Things will crash in unexpected ways and we want to make it easier for the developer to be able to dig through the Verona stack trace until they get to the actual error.

Sandboxing model fits with its expected usage, but we should also aim to simplify error handling to isolate the sandbox error from the rest of Verona (this is vague, I know). Just thinking out loud, don't mind this comment.

Reply:
Ah, I understand.  Yes, we can't guarantee ABI's aren't subtly broken in foreign code (though we can provide tooling to make it easier for anything that uses semantic versioning).  

Sandbox failure is an interesting topic that we don't yet have a complete story on.  We want to kill the process on contract violation.  We shouldn't kill the process when untrusted code misbehaves because doing so gives malicious code an easy DoS vector.  We also don't want to turn every sandbox call into something that returns an option type that needs checking between calls.  Unless we make every argument to sandboxed functions implicitly take an option type, so we can chain together calls, but need to check them when we return to Verona.  That might actually be a nice intermediate.  For example, the following C++ method:

```c++
Value *CreateRet(Value *);
```

Would be exposed in Verona as:

```verona
Value* | SandboxError CreateRet(Value* | SandboxError);
```

You could then pass around all of the `Value* | SandboxError` values from similar calls but when you wanted to store them anywhere you'd need to check if the sandbox had done a bad thing. If you pass the failure value into any of the wrapper functions, the compiler would insert code to skip the call.  

This would let you write a `when` clause that the compiler would turn into an asynchronous invocation that ran a bunch of commands in the sandbox asynchronously and returned a promise that would evaluate to `SandboxError` if the sandbox died.  If we catch the signal and preserve the error state within the sandbox, then we could get a stack trace (or even detach the child and hand if over to an interactive debugger) when the promise is accessed.

Reply:
I like the chain of optionals to only be checked at usage and the sandbox error state to be kept (not necessarily inside a sandbox object) until it's out of scope (whether or not it's consumed). 

This will be tricky to implement but it would be a really clean interface for sandbox ewrror handling.

Another way of DoS would be to have multiple failures with large error contexts that never go away. Say this case (in clearly no Verona code):

```
  Sandbox s;
  while (true) {
    when (resource L) {
       s(L).dispatch(); // or something
       // crashes, (s)'s error is kept, but not out of context
       s.tearDown(); // Doesn't remove the error
    }
    // all previous invocations of the external code accumulate their errors
    // they never go out of context, we never make use of them
    // How many of these can we keep in memory?
  }
```

So, perhaps we should also make sure that the same sandbox can only have one error context (the last one) and to be cleared once it's invoked again. What's not clear to me is how to differentiate them if they're running on different resources `L` (which could be a template or inherited parameter).

Reply:
Once a sandbox has hit an error, it *must* be torn down (or, at least, never invoked again).  We did that in CHERI-JNI because calling into something in an undefined state is... ill-advised.  

Your example wouldn't type check because `s` is a region and so the `when` clause would claim ownership of it and `s` effectively drops out of scope after the first loop iteration, so the second loop iteration is invalid code.  I'm not sure if I can transform it into something that typechecks and shows the problem.